content/match-types-spec.md

```diff
@@ -0,0 +1,399 @@
+It is however not possible to guarantee that property for *all* cases, since the existing implementation is not specified in the first place.
+
+## Proposed solution
+
+### Specification
+
+#### Preamble
+
+Some of the concepts mentioned here are defined in the existing Scala 3 specification draft.
+That draft can be found in the dotty repository at https://github.com/lampepfl/dotty/tree/main/docs/_spec.
+It is not rendered anywhere yet, though.
+
+Here are some of the relevant concepts that are perhaps lesser-known:
+
+* Chapter 3, section "Internal types": concrete v abstract syntax of types.
+* Chapter 3, section "Base Type": the `baseType` function.
+* Chapter 3, section "Definitions": whether a type is concrete or abstract (unrelated to the concrete or abstract *syntax*).
+
+#### Syntax
+
+Syntactically, nothing changes.
+The way that a pattern is parsed and type captures identified is kept as is.
+
+Once type captures are identified, we can represent the *abstract* syntax of a pattern as follows:
+
+```
+// Top-level pattern
+MatchTypePattern ::= TypeWithoutCapture
+                   | MatchTypeAppliedPattern
+
+// A type that does not contain any capture, such as `Int` or `List[String]`
+TypeWithoutCapture ::= Type // `Type` is from the "Internal types" section of the spec
+
+// Applied type pattern with at least one capture, such as `List[Seq[t]]` or `*:[Int, t]`
+MatchTypeAppliedPattern ::= TyconWithoutCapture ‘[‘ MatchTypeSubPattern { ‘,‘ MatchTypeSubPattern } ‘]‘
+
+// The type constructor of a MatchTypeAppliedPattern never contains any captures
+TyconWithoutCapture ::= Type
+
+// Type arguments can be captures, types without captures, or nested applied patterns
+MatchTypeSubPattern ::= TypeCapture
+                      | TypeWithoutCapture
+                      | MatchTypeAppliedPattern
+
+TypeCapture ::= NamedTypeCapture    // e.g., `t`
+              | WildcardTypeCapture // `_` (with inferred bounds)
+```
+
+In the concrete syntax, `MatchTypeAppliedPattern`s can take the form of `InfixType`s.
+A common example is `case h *: t =>`, which is desugared into `case *:[h, t] =>`.
+
+The cases `MatchTypeAppliedPattern` are only chosen if they contain at least one `TypeCapture`.
+Otherwise, they are considered `TypeWithoutCapture` instead.
+Each named capture appears exactly once.
+
+#### Legal patterns
+
+A `MatchTypePattern` is legal if and only if one of the following is true:
+
+* It is a `TypeWithoutCapture`, or
+* It is a `MatchTypeAppliedPattern` with a legal `TyconWithoutCapture` and each of its arguments is either:
+  * A `TypeCapture`, or
+  * A `TypeWithoutCapture`, or
+  * The type constructor is *covariant* in that parameter, and the argument is recursively a legal `MatchTypeAppliedPattern`.
+
+A `TyconWithoutCapture` is legal if one of the following is true:
+
+* It is a *class* type constructor, or
+* It is the `scala.compiletime.ops.int.S` type constructor, or
```

Comment:
shouldn't this also cover `compiletime.ops.long.S` [Source code](https://github.com/lampepfl/dotty/blob/3.3.0/library/src/scala/compiletime/ops/long.scala#L18)

Reply:
Perhaps, if we ignore the fact that [scala.compiletime.ops.long.S doesn't work](https://github.com/lampepfl/dotty/issues/18342) :man_shrugging:

Reply:
`long.S` was [officially deprecated](https://github.com/lampepfl/dotty/pull/18426), so this is not an issue anymore.